### ISSUE TITLE
ActiveRecord time not included in request logs when `shard_resolver`  is used

### ISSUE THREAD

**Issue by pospieszynski** [rails/rails#48687]
### Steps to reproduce
1. Create Rails 7.0.4.3 app in `api_only` mode
2. Configure database for multi-shards as described [here](https://edgeguides.rubyonrails.org/active_record_multiple_databases.html)
3. Add sample `shard_resolver` to `application.rb` file.
3. Implement sample endpoint that calls any db query and observe to last line of logs processing.
4. Put the code below to the initializer.

```
ActiveSupport::Notifications.subscribe "process_action.action_controller" do |event|
  Rails.logger.info "Duration: #{event.duration}; DB: #{event.payload[:db_runtime].to_i}"
  # Notice that event.payload[:db_runtime] is always nil here
end
```

### Expected behaviour
The log line that summarises the request completion should contain `ActiveRecord` time. Also, the instrumentation hook for `process_action.action_controller` event should have `event.payload[:db_runtime]` not being nil. Please see SS below:

<img width="540" alt="image" src="https://github.com/rails/rails/assets/15077823/29574765-0dfd-4d14-b044-d4fb1528ffc4">

### Actual behavior
Rails doesn't seem to correctly keep track of database time throughout the request lifecycle for multi-db setup resulting in missing ActiveRecord time in logs. Please see SS below:

<img width="388" alt="image" src="https://github.com/rails/rails/assets/15077823/722cf567-3e1e-4651-884f-96853c2aa35a">

### System configuration
**Rails version**: 7.0.4.3

**Ruby version**: 3.0.2

**Comment by eileencodes** [comment]
Hi @pospieszynski thanks for opening an issue. I wasn't able to reproduce this based off the description, can you provide a demo application that reproduces the behavior you're seeing? Thanks!

**Comment by pospieszynski** [comment]
Hi @eileencodes , Here is the sample app https://github.com/pospieszynski/testapp

It seems as though when you add a **shard_resolver** the logs in question disappear. Looks like Rails is not using the connection right to calculate the time. You can comment the `config.active_record.shard_resolver = -> (request) { :dogs }` and the proper logs are back. 🤔

**Comment by pospieszynski** [comment]
Thank you @eileencodes ! Do you have some idea when we can expect that to be included in the release? Not sure if it makes sense for me to wait for a while or just install it from Github source.

**Comment by eileencodes** [comment]
We don't have a scheduled release cadence but this will be in the next 7.x bug fix release (7.0.7) and the release of 7.1.0. You can find this infomrat
